### ISSUE TITLE
Support BigQuery as a datasource

### ISSUE THREAD

**Issue by tazmaniax** [appsmithorg/appsmith#3129]
## Summary

Support BigQuery as a datasource

## Motivation

A typical data pipeline is to replicate Firestore document inserts and updates in realtime to BigQuery using the [Firestore Export to BigQuery extension](https://firebase.google.com/products/extensions/firestore-bigquery-export). This then enables richer SQL queries against the data then would otherwise be possible directly through the Firestore API and this brings the Firebase data experience closer to a typical RDBMS.

![Firestore BigQuery Data Pipeline](https://user-images.githubusercontent.com/1079943/108625129-075a8080-7441-11eb-97ea-6b620e0aac07.jpg)

## Additional Context

[Google BigQuery API](https://cloud.google.com/bigquery/docs/reference/libraries)

**Comment by mohit-ula** [comment]
Looking forward to this integration.

**Comment by ajinkyakulkarni** [comment]
@mohit-ula Would it be possible to get in touch with you through our [Discord](https://discord.gg/rBTTVJp) server? Would like to understand your requirements better.

**Comment by qxxxoDev** [comment]
Yup, we need it so much and looking forward to seeing this integration.

**Comment by atedir** [comment]
Yup, we need it

**Comment by pjpimentel** [comment]
we need it here too!

**Comment by snegireff** [comment]
Yup, we need it

**Comment by netsatan** [comment]
Hey there. Looks very useful to me as well, if you need any help with implementing or testing that, let me know!

**Comment by jerrydeng** [comment]
We need it as well.

**Comment by JonasJacobsUserspace** [comment]
We also need it in the company. We use MySQL for updates and Bigquery for logging and reporting.

**Comment by ma-sada** [comment]
My company does development work for a wide variety of clients who tend to be heavy BigQuery users. Looking for the ideal low-code solution to recommend -- BQ integration is essential. Happy to discuss these use cases if it would help escalate this issue.

**Comment by mousetree** [comment]
Reading from BigQuery is the only thing missing for us to move from Retool to Appsmith

**Comment by tskippervold** [comment]
We use Appsmith as a back office tool where we also display data from our Google Cloud SQL database.
However, we have additional data in Mixpanel, such as user last seen, user device details, and so on.

Currently Mixpanel supports hourly exports to BigQuery, and this lets us query user data from big query.

It would be really nice to be able to pull in data from BigQuery in AppSmith so we can enrich our back office with this data.

**Comment by yan-hic** [comment]
Not sure if you folks know but you can refresh BQ data in Google Sheets, which you can then fetch in appsmith.
See https://cloud.google.com/bigquery/docs/connected-sheets

Note you cannot pull data from the Connected Sheet itself as it's not a true worksheet. You must create an `Extract` (separate sheet) which can then be pulled with the GSheets API.
Make sure to schedule the refresh i.e. every hr or 5 min - it doesn't cost anything if data hasn't changed (cached).

Add'l benefits: you can apply any other transformation in GSheets e.g pivot or any calculation/aggregation and pull in appsmith.

**Comment by Nikhil-Nandagopal** [comment]
@yan-hic that's a really cool workaround! Thanks for sharing that with the community

**Comment by rohan-arthur** [comment]
We haven't gotten around to this yet, but meanwhile here's another workaround!
[Hasura](hasura.io) now supports integration with BigQuery: https://hasura.io/docs/latest/databases/bigquery/getting-started/cloud/

What this means is, you can integrate Hasura with BigQuery, and the added advantage with this is that Hasura exposes all of that data via GraphQL. We know that Appsmith [plays very well](https://docs.appsmith.com/reference/datasources/graphql) with GraphQL.

**Comment by yan-hic** [comment]
Caveat: Hasura BigQuery integration is read-only. It doesn't support writing to BQ, so no insert/update/delete. 
Granted, Connected Sheets as mentioned before neither but you can have Appsmith write the GSheets, and read Sheets from BigQuery (external table)

**Comment by amirdarwish** [comment]
Using sheets is a limited workaround for limited use cases, not a real solution.

**Comment by yan-hic** [comment]
Here another great workaround to read from - and write to ! - BigQuery. Use `supabase` as middleman !
Why ? `supabase` exposes wrappers and a recent addition is the connection to BigQuery - not documented well everywhere, just follow steps 3 onwards at  https://github.com/supabase/wrappers/tree/main/wrappers/src/fdw/bigquery_fdw
This will allow you to read/write from/to BigQuery from within PostGreSQL.

Then, instead of using the PG integration from Appsmith (which has still issues re pooling), use instead the `supabase` js lib (must be added to your project, just one click), and write your own fetching function to use in table source,...


There are many other wrappers available that can be useful in supabase or indirectly in Appsmith !

**Comment by sebak2003** [comment]
Please this is so much needed! APPSMITH you need to add this! Please.

In the meantime, what is the least expensive and easiest existing workaround to implement?

**Comment by GreenFlux** [comment]
as a workaround, you can connect Hasura to BigQuery, and expose it as a GraphQL endpoint. Then you can use GraphQL queries from Appsmith to Hasura to access your BigQuery data. 
https://hasura.io/blog/tutorial-get-instant-graphql-apis-on-bigquery-with-hasura/

Thanks @rohan-arthur, for figuring out this workaround.
